pkg/controller/podprobemarker/pod_probe_marker_controller.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	// watch for changes to PodProbeMarker
+	if err = c.Watch(&source.Kind{Type: &appsv1alpha1.PodProbeMarker{}}, &enqueueRequestForPodProbeMarker{}); err != nil {
+		return err
+	}
+
+	// watch for changes to NodePodProbe
+	if err = c.Watch(&source.Kind{Type: &appsv1alpha1.NodePodProbe{}}, &enqueueRequestForNodePodProbe{}); err != nil {
+		return err
+	}
+
+	// watch for changes to pod
+	if err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &enqueueRequestForPod{reader: mgr.GetClient()}); err != nil {
+		return err
+	}
+
+	// watch for changes to node
+	if err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &enqueueRequestForNode{Reader: mgr.GetClient()}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcilePodProbeMarker{}
+
+// ReconcilePodProbeMarker reconciles a PodProbeMarker object
+type ReconcilePodProbeMarker struct {
+	client.Client
+	scheme *runtime.Scheme
+
+	finder *controllerfinder.ControllerFinder
+}
+
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=podprobemarkers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=podprobemarkers/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=nodepodprobes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=nodepodprobes/status,verbs=get;update;patch
+
+// Reconcile reads that state of the cluster for a PodProbeMarker object and makes changes based on the state read
+// and what is in the PodProbeMarker.Spec
+func (r *ReconcilePodProbeMarker) Reconcile(_ context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var err error
+	if strings.Contains(req.Name, ReconPodProbeMarker) {
+		err = r.syncPodProbeMarker(req.Namespace, strings.Split(req.Name, "#")[1])
+	} else {
+		err = r.syncNodePodProbe(strings.Split(req.Name, "#")[1])
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ReconcilePodProbeMarker) syncPodProbeMarker(ns, name string) error {
+	// Fetch the PodProbeMarker instance
+	ppm := &appsv1alpha1.PodProbeMarker{}
+	err := r.Get(context.TODO(), client.ObjectKey{Namespace: ns, Name: name}, ppm)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return nil
+		}
+		// Error reading the object - requeue the request.
+		return err
+	}
+	pods, err := r.getMatchingPods(ppm)
+	if err != nil {
+		klog.Errorf("PodProbeMarker ppm(%s/%s) list pods failed: %s", ppm.Namespace, ppm.Name, err.Error())
+		return err
+	}
+	// remove podProbe from NodePodProbe.Spec
+	if !ppm.DeletionTimestamp.IsZero() {
+		return r.handlerPodProbeMarkerFinalizer(ppm, pods)
+	}
+	// add finalizer
+	if !controllerutil.ContainsFinalizer(ppm, PodProbeMarkerFinalizer) {
+		err = util.UpdateFinalizer(r.Client, ppm, util.AddFinalizerOpType, PodProbeMarkerFinalizer)
```

Comment:
can we add the finalizer in webhook , so that we don't need to update podmarker itself

Reply:
There is currently no logic for mutating webhook, is it a bit heavy for only finalizer?